[리뷰요청] 전체 스타일 변경 -  input Validation, 전체 맵 스타일링 함수, 변경상태 추가

Eunsol0410 opened the pull request:
## 추가된 주요 기능
- useWholeStyle Hook의 input 스타일 검증
- useWholeStyle Hook의 flag를 이용하여 전체 맵스타일링 실행 추가 
- import JSON 모달 창에 useWholeStyle Hook 함수 적용하여 테스트 완료
- SET_WHOLE 액션에서 기능/세부 유형의 isChanged 상태 처리

## 기타 사항
- 위 함수를 구현하면서 영향을 받는 기존 함수들에도 일부 수정이 있었습니다.
- 전체 스타일 변경이 걱정했던 것보다 랜더링 시간이 오래 걸리지 않아서 다행이었습니다.
- 제가 이슈 번호를 잘못 적은 부분이 있어서 죄송합니다....ㅜㅜㅜ 덤벙거리는거 들켰네요😅 push 한 뒤에 PR하려고 할때 알아가지구.. 되돌리지 못했어요 
- 맵 스토어에서 분리하는건 제가 혼자 했다가 다른 분들과 의존성이 생겨 큰 변화를 일으킬 것 같다는 생각이 들어서 남겨두었습니다.

## 관련 이슈
- weight가 기본값이 0으로 되어있어서 기존 테두리나 라인이 다 없어지는 현상이 발생했습니다. 😢

Eunsol0410 committed:
[refactor] 불필요한 파라미터 제거, 파일 위치 및 이름 조정

- event 객체 전달 받을 필요 없어서 파라미터에서 제거함
- useModalStatus가 다른 모달창과 공통으로 사용되기 힘들었어서, Import창에 맞춰서 사용하기 위해 파일명과 위치 조정함

Eunsol0410 committed:
[fix] Import 클릭 후 처리

- 지도 가져오기 클릭 시 해당 입력값을 JSON으로 파싱하고 상태업데이트 처리하도록 모달창 Hook으로 이동하였음
- 잘못된 입력에 대해서 사용자에게 전달하기 위해서 inputStatus 상태를 추가하였음
- 파싱 에러를 추출하기 위해서 try catch 문을 사용하였음. 에러 발생하면 inputStatus false로 변경하도록 함

Eunsol0410 committed:
[feat] 스타일 검증 함수 구현

- import, export url 등을 통해서 전달되는 스타일 입력 값을 interface와 비교하는 함수를 구현함
- 위 함수에 필요한 타입이 있어서 공통 타입에서 분리함
- import 관련 Hook에서 validation 실패하면 에러 발생해서 catch문으로 들어가도록 하였음

Eunsol0410 committed:
[feat] 입력 상태 View에 반영

- 입력값 잘못된 경우 디바운스 방식으로 에러메시지 출력되도록 하였음
- 경고 메시지 붉은색 필요해서 theme 색상에 추가하였음

Eunsol0410 committed:
[fix] hsl -> hex 코드로 변경

- validation에서 색상 hex 코드 입력 검증으로 수정함

Eunsol0410 committed:
[fix] subElement 없는지 체크하는 조건 제거

- 라벨 아이콘의 경우는 subElement가 없기 때문에 해당 조건 뺐음

Eunsol0410 committed:
[feat] [#69] 전체 상태 변경시 실행되는 맵 스타일링 함수 구현

- 스토어 별 상태를 파라미터로 전달해서 실행하는 함수로 만들었음
- map을 스토어에서 분리시키기다가 이슈 발생할 것 같아서 리팩토링 전 버전으로 우선 만들었음

Eunsol0410 committed:
[feat] [#69] flag를 이용한 전체 맵스타일링 적용

- flag를 사용해서 true가 되면 전체 스타일 스토어의 맵 스타일링 함수 실행되도록 하였음

Eunsol0410 committed:
[fix] [#66] 모달창 컴포넌트 삭제를 flag 전환 이후로 수정

- 모달 창 컴포넌트 사라짐에 따라서 flag 상태 자체가 사라져서 맵스타일링 함수가 실행되지 못하는 이슈가 있었음
- flag를 모달창 상태 관리 Hook에 전달하여 flag가 false로 전환될 때 모달창도 닫힘 상태로 바뀌도록 수정하였음

Eunsol0410 committed:
[fix] hex 코드 대소문자 구분 없도록 수정

Eunsol0410 committed:
[fix] 전체 상태 변경 스타일 isChanged 체크 추가, 스타일 combine 함수 if문 및 중복 요소 제거

- input 스타일 속성이 포함되는 경우에만 변화 상태 체크하도록 하였음
- combine 내에서 이미 defaultStyle 갖고 있어서 checkStyleIsChanged 안에서 다시 불러오지 않도록 파라미터로 전달받을 수 있게 수정함
- switch 문이 더 명시적이고 중복되는 함수 구조도 제거될 것 같아서 리팩토링 하였음

Eunsol0410 committed:
[fix] 기능 유형의 스타일 변화 체크 추가

- 세부 유형 변화 체크와 동일하게 입력값에 포함되는 기능 유형 스타일만 변화되었는지 체크하도록 하였음
- 기본값은 isChanged가 false로 영향을 미치지 않아서 굳이 속성 삭제를 따로 하지 않았음

Eunsol0410 committed:
[fix] [#66] 전체 상태 변경 순서 설정

- 맵스타일링 순서에 따라서 변경되야할 상태 위해 기본값이 덮어져 변경되지 않는 문제가 있었음
- sort 함수 이용해서 isChanged가 false인 유형부터 맵스타일링 적용되도록 하였음

Eunsol0410 committed:
[fix] [#69] 전체 상태 변경 순서 설정

- 맵스타일링 순서에 따라서 변경되야할 상태 위해 기본값이 덮어져 변경되지 않는 문제가 있었음
- sort 함수 이용해서 isChanged가 false인 유형부터 맵스타일링 적용되도록 하였음

Eunsol0410 committed:
Merge branch 'feat/style' of https://github.com/Eunsol0410/Project08-A-Styled-Map-Admin-Tool into feat/style

qkrdmstlr3 commented on src/hooks/common/useWholeStyle.ts:
```diff
@@ -2,23 +2,77 @@ import mapboxgl from 'mapbox-gl';
+  const transit = useSelector<RootState>(
+    (state) => state.transit
+  ) as FeatureState;
+  const water = useSelector<RootState>((state) => state.water) as FeatureState;
```
구조분해할당으로 가져오는 것이 낫지 않을까요??
`
const { poi, road } = useSelector<RootState>(
    (state) => state
  ) as FeatureState;
`

Eunsol0410 replied:
네! 그게 좋을 것 같습니다

qkrdmstlr3 commented on src/hooks/common/useWholeStyle.ts:
```diff
@@ -2,23 +2,77 @@ import mapboxgl from 'mapbox-gl';
 
   const changeStyle = (inputStyle: WholeStyleActionPayload): void => {
+    setFlag(true);
     dispatch(setWholeStyle(inputStyle));
```
setFlag가 뒤로 가야 dispatch로 상태가 바뀐 뒤에 정확히 가져올수 있을 것 같아요.
잘은 모르지만 setFlag가 비동기적으로 작동하는 것으로 알고있는데, dispatch의 작업이 길어지게 된다면 상태가 바뀌기전에 flag가 바뀌어서 useEffect가 실행될수도 있지 않을까요??

Eunsol0410 replied:
테스트 해봤을 때는 dispatch 내부 함수 -> useEffect(poi, road 등..) -> useEffect(flag) 순으로 실행되는 걸 확인할 수 있었습니다.
랜더링 순서와 방식에 대해서 한번 찾아보고 공유해드리겠습니다!

qkrdmstlr3 replied:
감사합니다! 혹시나 싶어서 여쭤보았습니다!

Eunsol0410 replied:
이렇게 배울 수 있겠죠?? 😉

qkrdmstlr3 commented on src/hooks/common/useWholeStyle.ts:
```diff
@@ -2,23 +2,77 @@ import mapboxgl from 'mapbox-gl';
+      map,
+      feature: FeatureNameType.water,
+      featureState: water,
+    });
```
배열에 담아놓고 forEach로 반복을 도는 것은 어떨까요??

Eunsol0410 replied:
한번 시도해보겠습니다😁  안그래도 이게 동일한 작업이라.. 좀 그랬거든요!

qkrdmstlr3 commented on src/hooks/sidebar/useImportModalStatus.ts:
```diff
@@ -0,0 +1,62 @@
+import { useState, useEffect } from 'react';
+import useWholeStyle from '../common/useWholeStyle';
+import validateStyle from '../../utils/validateStyle';
```
hook에서 hook을 부르는 걸까요?
제 생각으로는 DetailType 컴포넌트 처럼 이 두개의 hook을 사용하는 쪽에서 useWholeStyle의 반환값을 받은 후 useImportModalStatus의 인자로 넘겨주는게 좋다고 생각합니다.(아니면 하나로 합치기?)

Eunsol0410 replied:
React 공식 페이지에도 hook이 hook을 사용하는 것에 대해서 설명하고 있어서, 별 문제가 없을 것 같다고 생각했는데... 다른 이유가 있을까요!? 궁금합니다!

qkrdmstlr3 replied:
제가 본 공식 페이지랑 같은지는 모르겠으나, Custom hook에서 react에서 제공하는 기본적인 hook을 호출하라고 되어있습니다. custom hook이 custom hook을 호출하는 것이 옳은지는 사실 저도 잘 모르겠습니다(본적이 없어서..)😅. (항상 custom hook은 컴포넌트에서만 호출하는 경우를 보아서 제가 부자연스럽게 느꼈을 수도 있습니다!). 단지 아까 hook이 어느 컴포넌트에서 사용되었는지를 찾았을 때 조금 혼란을 느껴서 드린 의견입니다!

Eunsol0410 replied:
아! 다시 확인해보니 제가 문서에 나온 예시를 잘못 이해했던 것 같습니다. ㅎㅎㅎ
[팁: Hook에서 Hook으로 정보 전달하기](https://ko.reactjs.org/docs/hooks-custom.html)

hook의 프로퍼티로 전달하라는 이야기였네요...🤣 같이 수정해서 커밋해보도록 하겠습니다.

qkrdmstlr3 commented on src/store/style/getReducer.ts:
```diff
@@ -81,11 +83,16 @@ export default function getReducer(IDX: number): ReducerType {
           });
+          updateStyle[subFeature] = {
+            ...newFeature,
+            isChanged: checkFeatureIsChanged(newFeature),
```
이거는 import시 변화를 감지하려고 하는 것이지요?

Eunsol0410 replied:
네네!! 맞아요 입력값과 기본값이 합쳐진 경우에 꼭 isChanged가 true라고 보긴 어려울 것 같아서요~

qkrdmstlr3 commented on src/hooks/sidebar/useImportModalStatus.ts:
```diff
@@ -0,0 +1,62 @@
+
+  const onClickOK = (inputText: string) => {
+    try {
+      if (!inputStatus) return;
```
이 함수를 통해서 json으로 입력 / url으로 파싱된 json / 테마가 모두 변환된다고 볼 수 있겠죠????

Eunsol0410 replied:
음.. 이건 모달창 닫기 관련된 기능이 같이 포함되어 있어서, 다른 것도 Hook을 만들고 useWholeStyle을 불러서 사용하는 방식을 생각했습니다!! 한번 해봐야 알 것 같기는 하네요 😁

qkrdmstlr3 replied:
음... 아까위에서 드린 의견인데, useWholeStyle이 다른곳에서도 사용되게 된다면, hook이 hook을 부르는 것이 조금 더 편할 것 같습니다!

Eunsol0410 replied:
음... 바꾸려고 했었는데...! 이부분도 다시 한번 찾아볼게요!

qkrdmstlr3 replied:
연정님이랑도 비슷한 문제로 이야기 중이었습니다.내일 다같이 한번 정리해봐도 좋을 것 같아요! custom hook과 react hook을 어떻게 정의하느냐가 문제겠네요

qkrdmstlr3 commented on src/hooks/sidebar/useImportModalStatus.ts:
```diff
@@ -0,0 +1,62 @@
+    if (inputStatus) return;
+    const timer = setTimeout(() => {
+      setInputStatus(true);
+    }, Delay);
```
inpustStatus를 통해서 성공 / 실패를 표기하는 것으로 보이는데 delay는 어떤 용도로 사용되나요?

Eunsol0410 replied:
디바운스 적용해서 '지도 불러오기' 버튼이 일시적으로 실패 메시지로 전환되도록 하였습니다.
그래서 딜레이 시간 이후에 다시 '지도불러오기'로 전환되도록 상태를 전환시켰습니다. 그래야 버튼이 활성화되어서요~ 
![image](https://user-images.githubusercontent.com/41146374/101283559-78b16c80-381e-11eb-9491-30c6113ff9c8.png)
위 상태에서 2초 지나면 다시 지도 불러오기로 바뀝니다!

qkrdmstlr3 replied:
ㅇㅎ 감사합니다!

qkrdmstlr3 commented on src/store/style/getReducer.ts:
```diff
@@ -48,7 +48,9 @@ export default function getReducer(IDX: number): ReducerType {
 
-        style.isChanged = checkStyleIsChanged(action.payload);
+        const defaultStyle = getDefaultStyle(action.payload);
+        style.isChanged = checkStyleIsChanged({ defaultStyle, style });
```
오 한방에 가져올 수 있었네요 ㅋㅋ 👍

Eunsol0410 replied:
네!!! ㅎㅎㅎ 👍

qkrdmstlr3 commented on src/utils/validateStyle.ts:
```diff
@@ -0,0 +1,119 @@
+function checkElement(
+  input: ElementActionPayload,
+  origin: StyleElementPropsType
+): boolean {
```
주는 쪽에서는 subFeatureStyle, layer로 주고 받는 쪽에서는 input, origin으로 받는데 객체로 주고 받는 것은 어떨까요?

Eunsol0410 replied:
그럼 layer가 조금 명시적이지 않은 것 같아서, originStyle?로 바꾸고 은식님이 제시한 방식으로 하는게 어떨까요? 네이밍은 항상 어렵네요 😂
originSubFeatureStyle은 너무 긴 것 같아서....

qkrdmstlr3 replied:
30자 이내라면 저는 길어도 될 것 같아요!

Eunsol0410 replied:
시도해보겠습니다!

qkrdmstlr3 commented on src/utils/validateStyle.ts:
```diff
@@ -0,0 +1,119 @@
+      key === StyleKeyType.visibility &&
+      typeof input.visibility !== 'boolean'
+    )
+      return false;
```
모두가 false로 반환하는 것을 봐서는 하나로 통합이 가능할까요? 혹시 조건별로 구분을 위한 것이라면 각 조건문을 변수로 분리후 or연산자로 한번에 처리하는 것도 좋은 것 같습니다!

Eunsol0410 replied:
너무 길어져서 이해하기 어렵지 않을까 싶어서 구분을 하였는데, 후자의 방식을 하면 불필요한 연산을 실행하는 것 같기도 해서요! 
전자의 방식으로 했을 때 그리 어렵지 않을 것 같나요? 궁금합니다! 저는 코드를 짠 사람의 입장에서 보게 되어서.. ㅎㅎㅎ

qkrdmstlr3 replied:
음 아무래도 각 조건이 어떤 조건들인지 파악하기가 힘들었습니다. 만약 변수로 따로 선언을 해준다고 해도 연산량 자체에 있어서는 추가적인 ||문 말고는 차이가 없을 것 같습니다. 변수명이 있다면 나중에 코드를 다른사람이 수정할 때도 좀 더 쉬울 것 같다는 생각입니다!

Eunsol0410 replied:
음 전체적인 통일성을 생각해서 위 함수에서도 적용하는 걸 고려해서 말씀하셨다고 생각했는데, 아닐까요?
```
    if (
      !(subElement in SubElementNameType) ||
      typeof subElementSytle !== 'object'
    ) {
      return false;
    }

    if (!checkStyle(subElementSytle)) return false;
```
이런 부분에도 동일한 패턴을 적용한다면 변수로 빼면 checkStyle은 불필요한 경우에도 실행될 것 같아서요~ 
comment 주신 부분에서만 말씀해주신 걸까요!?

qkrdmstlr3 replied:
흠...결국은 가독성이냐 효율성이냐의 문제네요.
일종의 취향으로 갈릴 수 있겠으나, 마스터님들이 맨날 js성능이 엄청 좋다고 하셔서 가독성 쪽이기는 합니다..

qkrdmstlr3 replied:
흠...결국은 가독성이냐 효율성이냐의 문제네요.
일종의 취향으로 갈릴 수 있겠으나, 마스터님들이 맨날 js성능이 엄청 좋다고 하셔서 가독성 쪽이기는 합니다..

Eunsol0410 replied:
그럼! 다른분들의 의견도 기다려보겠습니다!

qkrdmstlr3 commented on src/utils/validateStyle.ts:
```diff
@@ -0,0 +1,119 @@
+  const color = inputColor.replaceAll(' ', '');
+  const hslReg = /^#[0-9A-F]{6}$/i;
+  if (!color.match(hslReg)) return false;
+  return true;
```
hex검사로 알고있었는데 hslReg는 무엇인가요?

Eunsol0410 replied:
아!! 처음에 hsl로 했어서 정규식만 바꾸고 변수명을 안바꿨네요 ㅎㅎㅎㅎ

qkrdmstlr3 reviewed (approved):
수고하셨습니다!
색상 / 채도 / 명도가 절댓값으로 처리된다면, import / export시에 색상 값만 받으면 채/명도는 자연스럽게 넣을 수 있다고 생각합니다. 그렇게되면 유효성 검사나 url길이 등 많은 부분에서 좀 더 효율적이라고 생각하는데 어떠신가요?

Eunsol0410 committed:
[refactor] 전체 스타일링 Hook 리팩토링

- destructuring, for문 이용해서 함수 간결하게 수정함
- destructuring 부분은 준희님이 구현하셨던 Hook중에 유사한게 있어서 그걸 사용해도 좋을 것 같기도 함

Eunsol0410 committed:
[refactor] 오타 수정

gitgitWi reviewed (approved):
고생하셨습니다
양이 엄청나네요 ㅎㅎ 지금 정신이 좀 맑지가 않아서 점심먹으면서 한번더 천천히 봐야할 거 같아요..
